files/archive.py

```diff
@@ -0,0 +1,385 @@
+                            arcfile.write(path, path[len(arcroot):])
+                        else:
+                            arcfile.add(path, path[len(arcroot):], recursive=False)
+
+                        successes.append(path)
+
+            except Exception:
+                e = get_exception()
+                return module.fail_json(msg='Error when writing %s archive at %s: %s' % (compression == 'zip' and 'zip' or ('tar.' + compression), dest, str(e)))
+
+            if arcfile:
+                arcfile.close()
+                state = 'archive'
+
+            if len(errors) > 0:
+                module.fail_json(msg='Errors when writing archive at %s: %s' % (dest, '; '.join(errors)))
+
+        if state in ['archive', 'incomplete'] and remove:
+            for path in successes:
+                try:
+                    if os.path.isdir(path):
+                        shutil.rmtree(path)
+                    else:
+                        os.remove(path)
+                except OSError:
+                    e = get_exception()
+                    errors.append(path)
+
+            if len(errors) > 0:
+                module.fail_json(dest=dest, msg='Error deleting some source files: ' + str(e), files=errors)
+
+        # Rudimentary check: If size changed then file changed. Not perfect, but easy.
+        if os.path.getsize(dest) != size:
+            changed = True
+
+        if len(successes) and state != 'incomplete':
+            state = 'archive'
+
+    # Simple, single-file compression
+    else:
+        path = expanded_paths[0]
+
+        # No source or compressed file
+        if not (os.path.exists(path) or os.path.lexists(dest)):
+            state = 'absent'
+
+        # if it already exists and the source file isn't there, consider this done
+        elif not os.path.lexists(path) and os.path.lexists(dest):
+            state = 'compress'
+
+        else:
+            if module.check_mode:
+                if not os.path.exists(dest):
+                    changed = True
+            else:
+                size = 0
+                f_in = f_out = arcfile = None
+
+                if os.path.lexists(dest):
+                    size = os.path.getsize(dest)
+
+                try:
+                    if compression == 'zip':
+                        arcfile = zipfile.ZipFile(dest, 'w', zipfile.ZIP_DEFLATED)
+                        arcfile.write(path, path[len(arcroot):])
+                        arcfile.close()
+                        state = 'archive' # because all zip files are archives
+
+                    else:
+                        f_in = open(path, 'rb')
+
+                        if compression == 'gz':
+                            f_out = gzip.open(dest, 'wb')
+                        elif compression == 'bz2':
+                            f_out = bz2.BZ2File(dest, 'wb')
+                        else:
+                            raise OSError("Invalid compression")
```

Comment:
This could be detected earlier and avoid the slightly odd use of OSError here (presumably to trigger the except handler below?).